### DISCUSSION TITLE
Language dependent testing features

### DISCUSSION THREAD

**Discussion by jrstrunk** [gleam-lang/gleam#3819] [Ideas & suggestions]
Hey all, I wanted to start a discussion about two relatively simple things I'd love to see in the way we test Gleam code that would require new language features (and could not be implemented by a cool new testing framework):

1. Some compiler-enforced way to show the tests a function is tested by the alongside function definition (so at a glance I can see how a function is being tested, for understanding the function or determining if it needs more test cases)
2. The ability to import private functions / constants from modules into test functions (library code has the @internal function, but application code is stuck with making everything public).

There are lots of ways to implement these two things and there are pros and cons of each, maybe we can all discuss this in the comments :)

**Comment by jrstrunk** [comment]
A simple thing that would satisfy both is to add a `@test` annotation that can be added to functions, which would specify a test function to receive the current function as part of it's scope. It could look something like:
```gleam
// In pair.gleam
// we'd have to first allow importing from the test/ dir, if not just allowed to be called in the @test annotation
import pair_test 

@test(with: pair_test.get_first_string_test)
fn first(pair: #(a, b)) -> a {
  pair.0
}
```
```gleam
// In pair_test.gleam

pub fn get_first_string_test() {
  first(#("Hi", "mom"))
  |> should.equal("Hi")
}
```

Because cyclical imports are not allowed, this would enforce not importing the modules (and their functions) to be tested into the test module, but instead lead authors to add the `@test` annotation to each function definition in order to test it, achieving item 2.

In tests that use multiple functions:
```gleam
// In pair.gleam
import pair_test

@test(with: pair_test.map_second_test)
@test(with: pair_test.get_second_test)
fn second(pair: #(a, b)) -> b {
  pair.0
}

@test(with: pair_test.map_second_test)
fn map_second(pair: #(a, b), to c) -> #(a, c) {
  #(pair.0, to)
}
```
```gleam
// In pair_test.gleam

pub fn map_second_test() {
  #("Hi", "Lucy")
  |> map_second("friend")
  |> second
  |> should.equal("friend")
}
```

This Gleam code would compile to Erlang or JS code that has the module imports reversed (so the test module imports the main one and not vice versa like in the Gleam code), and all the functions with the `@test` annotation are made public and added as unqualified imports to the test module. Compile time checks would need to be run for collisions between function names and emit an error if so.

This `@test` annotation could then later be expanded to do cool things that require more language features and consideration, similar to #3017:
```gleam
// In pair.gleam

@test(input: #("Hi", "Lucy"), output: "Hi")
fn first(pair: #(a, b)) -> a {
  pair.0
}
```

A real world example of library code using this might look like this (using the actual tests and functions in the tempo package):
```gleam
import datetime_test as test

/// Parses a datetime string in the format `YYYY-MM-DDThh:mm:ss.sTZD`,
/// `YYYYMMDDThhmmss.sTZD`, `YYYY-MM-DD hh:mm:ss.sTZD`,
/// `YYYYMMDD hhmmss.sTZD`, `YYYY-MM-DD`, `YYYY-M-D`, `YYYY/MM/DD`, 
/// `YYYY/M/D`, `YYYY.MM.DD`, `YYYY.M.D`, `YYYY_MM_DD`, `YYYY_M_D`, 
/// `YYYY MM DD`, `YYYY M D`, or `YYYYMMDD`.
/// 
/// ## Examples
/// 
/// ```gleam
/// datetime.from_string("20240613T230400.009+00:00")
/// // -> datetime.literal("2024-06-13T23:04:00.009Z")
/// ```
@test(with: test.negative_offset_from_string_test)
@test(with: test.positive_offset_from_string_test)
@test(with: test.z_offset_from_string_test)
@test(with: test.condensed_negative_offset_from_string_test)
@test(with: test.condensed_offset_from_string_test)
@test(with: test.char_offset_from_string_test)
@test(with: test.space_delim_from_string_test)
@test(with: test.naive_from_string_test)
@test(with: test.date_out_of_bounds_from_string_test)
@test(with: test.time_out_of_bounds_from_string_test)
pub fn from_string(
  datetime: String,
) -> Result(tempo.DateTime, tempo.DateTimeParseError) {
  let split_dt = case string.contains(datetime, "T") {
    True -> string.split(datetime, "T")
    False -> string.split(datetime, " ")
  }

  case split_dt {
    [date, time] -> {
      use date: tempo.Date <- result.try(
        date.from_string(date)
        |> result.map_error(fn(e) { tempo.DateTimeDateParseError(e) }),
      )

      use #(time, offset): #(String, String) <- result.try(
        split_time_and_offset(time),
      )

      use time: tempo.Time <- result.try(
        time.from_string(time)
        |> result.map_error(fn(e) { tempo.DateTimeTimeParseError(e) }),
      )
      use offset: tempo.Offset <- result.map(
        offset.from_string(offset)
        |> result.map_error(fn(e) { tempo.DateTimeOffsetParseError(e) }),
      )

      new(date, time, offset)
    }

    [date] ->
      date.from_string(date)
      |> result.map(new(_, tempo.time(0, 0, 0, 0, None, None), tempo.utc))
      |> result.map_error(fn(e) { tempo.DateTimeDateParseError(e) })

    _ -> Error(tempo.DateTimeInvalidFormat)
  }
}
```

Another cool thing this could enable is an in-code reminder to add tests later (maybe you thought of a niche test while in the middle of an implementation and want to write it down before you forget), emitting a warning but not actually crashing:
```gleam
@test(with: pair_test.map_second_test)
@test(with: todo as "get second with custom type test")
fn second(pair: #(a, b)) -> b {
  pair.0
}
```

**Reply by jrstrunk** [reply]
Alternatively, `@tested(by: pair_test.map_second_test)` on the main function and `@test(for: pair.map_second)` on the test function would be more explicit but requires more back-and-forth setup which is annoying. Plus I am not sure how you'd handle the Gleam import statements since this implies a cyclical import.

**Comment by jrstrunk** [comment]
A solution that would generally solve item 2 would be to add some way to have friend modules, and then make the test module a friend to the main module so it can import its private values.

**Reply by lpil** [reply]
Private functions will never be importable, that's the definition of private. If you wish to import them then you would use `@internal`.

> but application code is stuck with making everything public

Applications can use internal modules and definitions. There's no difference there.

**Reply by jrstrunk** [reply]
I don't think friend modules are a great idea, I just wanted to include it as a secondary comment in case it resonated somehow haha.

But in regards to the internal definitions: in application code usually there are custom, but totally separate modules that live in the same project -- for example even in a simple application I might have a `database.gleam` module, an `api.gleam` module, and a `user_interface.gleam`. If I don't want any other module directly calling the `database.get_connections` function, but I add `@internal` to it so I can test it, then all my other application code now gets auto complete suggestions for calling it when I start typing `database.get_friends` (just for example). This not only could lead to a bug, but also clutters the auto complete prompt. The larger the application gets and the more people collaborate on it, the larger this issue becomes.

Since library code does not live in a project alongside completely separate logic, it does not have this issue and the `@internal` annotation / internal module is fantastic.

**Reply by lpil** [reply]
Sorry, I don't know what you mean. There is no difference with `@internal` whether or not you have defined a function called `main` or not.

**Reply by jrstrunk** [reply]
`@internal` works the same, but is less effective at separating application code because it's scope is the entire project. Applications tend to have code with different concerns living in the same project, whereas Libraries tend to only have code that has the same overall concern in a project.

**Reply by lpil** [reply]
You can split an application into as many packages as you like. It's common for web applications to have at least 3 packages.

**Comment by giacomocavalieri** [comment]
I don't love the idea of exposing private functions to a test module, or testing the private bits of a module in general. 
In my experience testing the private stuff makes the tests harder to maintain and ties those to some implementation details that become harder to change; I like how gleam forces me to only test the public api of a module. 

I'm sure there's plenty of exceptions to this and there could be a use for test functions inside a module to test its private bits but I haven't really felt that need so far :)

**Reply by jrstrunk** [reply]
I completely and wholeheartedly agree with that approach when it comes to library code, but I think application code is really what needs it since often it has side effects that need to be worked around in tests.

For example if I have the following function in my application: 
```gleam
pub fn get_flight_stats(flight_id: String) {
  http.request(flight_data_provder_api <> flight_id)
  |> json.expecting(flight_stats_decoder)
}

fn flight_stats_decoder(dyn) {
  // decode flight stats
}
```

If I want to test that my flight_stats_decoder function works, I have a couple options:
1. Make it public and import it into a test function -- the easiest option and honestly what most people will do. This then could lead to a co-worker calling it for some other dynamic value somewhere else in the application, thinking it is meant as a general use function, leading to a bug. It also clutters auto-complete when calling the module, and makes the module's true intended public api very muddy. The larger a team and the larger an application, the worse this gets.
2. Write an api abstraction and modify the `get_flight_stats` function to take that abstraction as an argument -- this then makes all your application code more complex, and is a lot of effort to set up. Unless you are committed to full deterministic simulation testing or there is a nice http library with really easy mocking, most people will not do this because the work probably doesn't justify the benefit
3. Test in prod.

Most people will probably do 1 or 3, both of which are actively bad things in a project. If I had a way to expose just the decoder to the test module, then wow I could write up test for it in just a few mins and then move on and do more product things instead of orchestrating complex test abstractions.

Library code aims to be stable, so setting up test abstractions is often worth it.
Application (business logic) code adapts fast to changing requirements, so setting up test abstractions is much less often worth it until there is a very mature, stable product.

Testing code is actually a pretty big interest of mine (and is how I got started professionally), and I know you wrote the really cool birdie package for Gleam, so I'd love to hear what you (and anyone else too) think about these points!

**Reply by lpil** [reply]
You would use @internal in this specific case detailed here.

**Reply by jrstrunk** [reply]
In a library I would, but if this was an application that had, say user info logic in a `user` module right next to this pretend `flight` module, how could I stop a co-worker from importing `flight_info.flight_stats_decoder` into the `user` module and using it in some invalid way?

**Reply by lpil** [reply]
By making it `@internal`, indicating that it is private.

**Reply by jrstrunk** [reply]
Right, and you made a good point in the other thread: "You can split an application into as many packages as you like". 

This is an answer, but can start to feel clunky with big projects. I have a project that has 57 modules that are all related to the same application. Some are small modules, but there are at least 13 that need significant testing. I would have to create 13 different packages inside one repo, then if I made a change in one of those packages that had implications in others, I would have to either go into each project and run the tests manually myself (which would be very error prone), or setup my own test script.

Not horrible, not great. So far Gleam's DX has set a standard for being great.

I don't want a way to share private definitions precisely between modules, but I think it would be nice for testing. If we make testing hard on authors, then it won't be done properly, we'll have broken packages in the ecosystem, and people won't feel good about the things they build with Gleam.

Since Gleam is wonderfully just types and functions, it has a good foundation to be extremely, easily testable -- it's one of the things I like most. I don't want to add frivolous new features to the language, but I think if we support the DX for testing then the entire ecosystem will benefit.

**Reply by lpil** [reply]
In that situation you would use `@internal` as it lets you test the code while clearly marking the code as unusable elsewhere, and removing it from all documentation and editor tooling.

**Reply by jrstrunk** [reply]
When I mark a function as `@internal` I still see it show up as an auto-complete suggestion in VS code (with no indication that it is internal). Is that what you meant by hiding it from editor tooling?

**Reply by lpil** [reply]
Apologies, I was thinking of from other packages.

**Comment by lpil** [comment]
What happens if you forget to add a `@test` annotation to the other function? It seems like it would be easy to forget to make the pointer to some test modules, resulting in the test never getting run.

What do you do for tests that are not testing any specific function? There's no clear place to add the annotation such that the test function is turned into a test.

**Reply by jrstrunk** [reply]
In the implementation I suggested first, the test wouldn't compile. Since you don't import the function being tested into the test module, the test function then depends on the special `@test` annotation to have the function being called in it's scope. In the following test, the function `first` is undefined unless the `first` function definition has the `@test` annotation above it, pointing to this test.
```gleam
pub fn get_first_string_test() {
  first(#("Hi", "mom"))
  |> should.equal("Hi")
}
```

These annotations are not a way to manage how tests are run, but to manage what tests have in scope. Whatever test runner the user is using will determine how the tests are run.

**Reply by lpil** [reply]
I'll give you an example:

```gleam
// module `a`

// @test forgotten for `b.wobble` here
pub fn wibble() {
  1
}
```
```gleam
// module `b`
import a

pub fn wobble() {
  // do something with a.wibble
}
```
How would we avoid the problem of forgetting the annotation here?

**Reply by jrstrunk** [reply]
You wouldn't need the annotation in this test setup. This looks like how tests are currently set up in most Gleam projects. The annotations I suggested aren't meant to be a replacement for the current test setup, but a complement / alternative to it. The annotations aren't required, but provide a nice alternative way to organize tests. It does not change how tests are executed in any way. That still depends on the test runner.

**Reply by lpil** [reply]
The Gleam language does not have any testing capabilities built-in today, so any system would need to include a mechanism for defining tests.

If these annotations are not for defining tests then what is their purpose?

**Reply by jrstrunk** [reply]
They help organize the scope of functions that external testing frameworks have access to.

**Reply by lpil** [reply]
Concretely what would that do? What new behaviour would the language have?

**Reply by jrstrunk** [reply]
It would provide authors an efficient way to keep track of existing test cases and keep an eye out for missing test cases.

Only with a new annotation would there be a way for the compiler to enforce that if the user decides to use the `@test` annotations, then they have to list all tests that apply to that function, and only the tests that actually exist (naturally, because of imports and scoping), in an accessible place which is looked at often.

Lesser reason, but it would also provide a way to test a private function without having to open that function up to the entire project via the `@internal` annotation

**Reply by lpil** [reply]
What would it actually do in the language though?

**Reply by jrstrunk** [reply]
It wouldn't do anything else in the language: it's basically just an alternative, precise import syntax that has the semantics of testing.

If the `@test` annotation was implemented like this (or similar), we could later add more behaviors depending on the labeled args given, but that is really a different discussion.

**Reply by lpil** [reply]
What do you mean? It's an import syntax?

What specifically does it do? I don't understand your proposal.

**Reply by jrstrunk** [reply]
The statement `import gleam/otp` "pulls" code from the `otp` module into the current one. In Gleam test setups right now, we "pull" module code into a test module to test it. My proposal is to have an annotation that allows for a "push" import of code into a test module. This "push" based importing is useful because it allows us to see everywhere a function is used right next to the definition of the function.

I think it really only makes sense for testing, where tracking test coverage is very valuable -- which is why I suggested an annotation named `@test(with: _)`, and not something like `@push(to: _)` or `@expose(to: _)`:

The following annotation would import the `first` function into the scope of the `test.first_string_test` function. The `test` module would not need any `import` statement to receive the `first` function.
```gleam
@test(with: test.first_string_test)
pub fn first()
```

**Reply by lpil** [reply]
I see!

There are some significant downsides to this design:

- It breaks Gleam's "only one way to do things" rule.
- The origin of all non-prelude names within a module can no longer be determined just from the file, harming readability.
- It is an unusual seeming feature so it spends more of the language's "strangeness budget".
- The module dependency graph now has cycles, making build tool caching more challenging and having a negative impact on compile times.
- The programmer now needs to edit two files to write a test.

To be something we can consider adding the advantages would need to sufficiently compensate for these and result in a large net positive overall.

What would you say the advantages are of this addition?

**Reply by jrstrunk** [reply]
Ah yeah for some reason I thought we could get around the cyclical dependencies, but I guess not. Looks like the theoretical implementation should be something else. 

I know `test` is a reserved keyword, is there any documentation / discussion about what you are leaning towards for that keyword?

Do you have any thoughts about the idea to include some sort of reference to a function's tests alongside the function definition?
I would say the advantage of this as an idea is having test coverage right in front of you every time a function is looked it. This will help authors catch lapses in test coverage without having to rely on external tools and will also help open source library contributors understand what a function needs to hold up to before they try and make a change.

**Reply by lpil** [reply]
The best way to think about language design (or any system design really) is to think about problems and requirements. Once they are fleshed out you can start to work towards some solutions and evaluate them against those goals.

Starting with a solution puts you in a place where you don't have enough context to evaluate it, so it's hard to argue for it.

No particular plans today. Problems I am interested in exploring are test discovery without resorting to brittle naming conventions and runtime reflection, type safe setup and tear down, test parallelism, improving test assertion output, documentation example verification.

**Reply by jrstrunk** [reply]
That's a good way of framing. I guess to put it that way: a problem is that in a medium project I have lost track of my test coverage, and I am unsure of what test cases may be missing. 

For a concrete example: `gtempo v5` has 197 public functions and over 400 tests. The details of which functions have which coverage is very difficult to figure out manually. I tried my best to be diligent and write tests immediately after writing a function so I had its full context in my working memory, but that is error prone. If I got distracted while writing a function, I may have lost track of a test I wanted to add. For example, in `gtempo v4.3.0` I just straight missed parsing the `Z` offset in the `datetime.parse_any` function. I might have caught it before publishing that version if I had an easy to review list of tests that applied to each function. If I saw `parse_any_positive_offset_test`, `parse_any_negative_offset_test`, and `parse_any_no_offset_test` listed together near the `datetime.parse_any` function, then when I was doing a review before publishing I probably would've realized the `Z` case was missing. As it is though, it is very un-user-friendly to manually review test coverage when you have a decent amount of them. 

It seems the answer right now it to use some external tool (don't think it exists for Gleam yet) or organization process. Maybe that is good enough -- but `gtempo` is bigger than I planned on it being, and I didn't start off thinking I'd need detailed test organization: meaning I accidentally walked into tech debt that could've been avoided if the compiler was guiding me from the beginning.

I'm planning on spending my next few months of free time studying testing for critical software. If I learn something useful I'd like to think of how it might apply to Gleam. Gleam has so many wonderful things about it; I'd love to see the testing DX be one of those things :)

**Reply by lpil** [reply]
Wouldn't you want a traditional coverage tool in this case? Having just the annotation on the function wouldn't be sufficient as it wouldn't show you if you hit all the paths though the function so you'd still be lacking the information required to know you missed a case such as `Z` offset parsing.

Erlang's [xref](https://www.erlang.org/doc/apps/tools/xref.html) and [cover](https://www.erlang.org/doc/apps/tools/cover.html) could be a good place to start.

**Reply by jrstrunk** [reply]
In the past I've never used a coverage tool that was smart enough to provide clear, useful feedback. Maybe I just haven't used the good ones. I'll give xref and cover a shot and see how they are!

**Reply by jrstrunk** [reply]
Also just in my opinion I like having test cases presented in the forefront instead of hidden behind a separate command, because then developers are lead into thinking about them more often :)

**Comment by lpil** [comment]
It is sometimes useful to have setup and teardown around all the tests, or around groups of tests. Did you have any thoughts for how this might work?

**Reply by jrstrunk** [reply]
The implementation that I suggested first would not control the execution of the test, just what is in scope of the test. Whatever test runner you were using would determine how to do test setup and teardown.

Since these annotations really only have the semantics of testing (and nothing that really strictly makes them apply to tests), they could also be applied to any setup/teardown functions like so:
```gleam
@test(with: pair_test.setup_helper)
@test(with: pair_test.complex_test_1)
@test(with: pair_test.complex_test_2)
fn first(pair: #(a, b)) -> a {
  pair.0
}
```

**Reply by lpil** [reply]
The programmer would still need to define the setup and teardown, so a design would be needed for this.

What would that proposed syntax do?

My concern with that one is that it looks extremely similar to the syntax for saying that `pair_test.setup_helper` is the test for the `first` function. It would also mean that adding a label changes the meaning of the expression, which is not how Gleam works today.

**Reply by jrstrunk** [reply]
We might as well hop in a call hahaha, the annotation I suggested does not execute any code itself. The setup and teardown design comes from whatever test runner the user is using, like gleeunit. This is not a replacement for a separate test running library, it is just a way to manage the scope of functions a test function in any arbitrary test framework has access to.

**Reply by lpil** [reply]
If the language is going to have syntax for defining setup and tear-down for tests it will need to be fully specified and type safe, including the interface that test runners would interface with in order to use it.

**Comment by inoas** [comment]
I'd love to have doctests:

- lsp to run them on hover or at least validate the example code to be actual gleam code
- something that the testing framework can hook into to run them

And I'd like to have some coverage info from gleam tooling if a larger testing feature is added from to gleam tooling

**Reply by jrstrunk** [reply]
Me too! Though that is different and much harder to implement than what I outlined above. Though if we were to expand what I suggested, what do you think of using the `@test(input: _, output: _)` annotations (or similar) for a doctests like experience instead of having the tests literally in the doc comment? I personally find it strange to have code in a comment be executed haha.

**Reply by inoas** [reply]
doctests make certain that the examples are not out of date
they also offer basic/minimum coverage

**Reply by jrstrunk** [reply]
Yeah and that's great! We could also generate docs from the `@test` annotation, so I think they bring the same value. I was just thinking about syntax and semantics mostly. 

But also, in functions that have a lot of tests, we may not want to display the tests as part of the main documentation snippet. It would be nice to have a way to separate the most value-bringing examples and the repetitive tests. I think allowing for formatting in the doc comment (like we currently have), plus having a second syntax for a test annotation would bring the best feature set.

**Reply by lpil** [reply]
Tests are not useful primary documentation, they communicate information very poorly and are too numerous and too verbose. Collecting all the tests to present them as the documentation is not an option.

**Reply by jrstrunk** [reply]
Totally agree, not primary documentation. They are useful to present in some niche view though: I can't count how many times I've learned how a library works from reading over the tests in lieu of incomplete documentation.
